### PR TITLE
StudentFeedbackSubmissionEditPage: Disable text area for closed feedback session #7797

### LINKED ISSUE #7797: StudentFeedbackSubmissionEditPage: Disable text area for closed feedback session
Text areas on StudentFeedbackSubmissionEditPage are not disabled and are still editable for a closed feedback session. They should be disabled since feedback can no longer be edited.

### PR DESCRIPTION
**Pull request by amarlearning**
Fixes #7797 

Looks like there wasn't any `text-area`, it is `div` which is used to collect the response from students. _[screenshot]_
screenshot from 2017-07-22 03-13-46

Attribute `disabled` is already to the `div`. since `div` don't support the `disabled` attribute that's why the `response text area` is enabled even when the session is closed.

**Outline of Solution**

- New Html attribute `data-is-session-open` is added to check if the session is open or closed in the `feedbackQuestionTextSubmissionFormTemplate.html` template.

- Removed the `DISABLED` property from `FeedbackTextQuestionDetails.java` as `div` don't support the `disabled` attribute.

- New variable `isSubmissionNotAllowed` is added to `feedbackSubmissionsEdit.es6` to check for the value of `data-is-session-open` data-attribute and update the `readonly`config property of TinyMCE for each question.

### CODE REVIEW

**Comment by amarlearning** [comment]
Ready for review

**Comment by HirdayGupta** [comment]
@amarlearning some comments:

- Good catch on the `disabled` attribute not being applicable to the `div`. However, I don't fully agree with your alternative.
- The problem is that the `.disabled-div` approach seems a little too fancy and hackish. Everything from the CSS properties it sets to the div on which it is placed seems irregular in my opinion. (The col-sm-10 div is very non specific)
- The fields that we are trying to set as disabled here are `tinyMCE` text editor fields. As such, for the `tinyMCE` editors, we always disable them during their initialisation in the javascript by setting its readonly property to true. This would be more concise and less hackish than the current CSS approach.
- A way in which you could do this is by embedding a data attribute in the HTML template for the text feedback question (Hint: its the` feedbackQuestionTextSubmissionFormTemplate.html`) e.g.: `data-is-session-open-for-submission`. You can follow the approach used to embed other data attributes like `data-length-text-id`. (While you're at it you can also see what to do about the `${disabled}` property that gets embedded there as you pointed out that it is useless).
- After the data is embedded, all you have to do is read it in the `feedbackSubmissionEdit.es6` file and change the readonly property of the `tinyMCE` editor which is initialised for all text type questions according to whether or not the session is open for submission.

**Comment by amarlearning** [comment]
@HirdayGupta Thanks for the help.
But I was thinking of replacing the `div` tag to `textarea` tag. `tinyMCE` and `disabled` attribute both work fine with `textarea`. 

What do you think of this approach?

**Comment by HirdayGupta** [comment]
@amarlearning `tinyMCE` seems to work fine with both divs and textareas. I'm not sure about the merits of one over the other. Maybe one of the senior devs could advise us about this. (But I can tell you that the `readonly` property works fine with both as well).

Regarding the `disabled` attribute however, I would still suggest using `tinyMCE`'s readonly property as that is the approach I have seen through other parts of the system. As I see it, once you hand control of a div/textarea to tinyMCE, you should let it handle the customisations you want rather than trying to mess with the attributes with yourself (tinyMCE has its own properties like `contenteditable` which it sets to the appropriate values if you set the readonly flag on init). If you want an example of setting tinyMCE's readonly property you can refer to [this line](https://github.com/TEAMMATES/teammates/blob/master/src/main/webapp/dev/js/main/instructorFeedbackEdit.es6#L247) which disables editing of the Session Instructions Rich text field on the InstructorFeedbackEditPage.

**Comment by amarlearning** [comment]
@HirdayGupta I was going through the TinyMCE docs and found that `inline only works with content that is a block element (e.g. div, h1)`.

`textarea` does not support inline editing. I think this is the reason why we are using `div` instead of `textarea`. _[[Reference link](https://www.tinymce.com/docs/get-started/use-tinymce-inline/#enablinginlineeditingmode)]_

Thanks for suggesting me [this](https://github.com/TEAMMATES/teammates/pull/7811#issuecomment-317178590) approach. :+1:

**Comment by HirdayGupta** [comment]
>  I was going through the TinyMCE docs and found that inline only works with content that is a block element (e.g. div, h1).

@amarlearning Ah, that makes sense. Thanks for reading and clearing that up 👍 

> Thanks for suggesting me this approach 👍 Will update my PR soon. 

Sure! 😃 Go ahead 👍

**Comment by amarlearning** [comment]
Ready for review

File: src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6 [hunk 1]
```diff
@@ -963,10 +963,12 @@ $(document).ready(() => {
     if (typeof richTextEditorBuilder !== 'undefined') {
         $.each(textFields, (i, textField) => {
             const id = $(textField).attr('id');
+            const isSubmissionNotAllowed = textField.dataset.isSessionOpen === 'false';
```
Comment (HirdayGupta) [inline src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6]: I think it would be better if you used the jQuery approach for retrieving data. It would be something like this: 
```javascript
$(textField).data('is-session-open')
```

And then, if you pass down `is-session-closed` instead, as I suggested in my comment below, you would not even have to check for equality with a boolean string as jQuery will automatically convert to boolean for you.

Note: I would also recommend changing the variable name to `isSessionOpen` or `isSessionClosed` (depending on the approach we go for) for consistency as `allowed/notAllowed` seems more like a user-role permissions related variable.

File: src/main/java/teammates/common/datatransfer/questions/FeedbackTextQuestionDetails.java [hunk 2]
```diff
@@ -57,7 +57,7 @@ public String getQuestionWithExistingResponseSubmissionFormHtml(boolean sessionI
             int responseIdx, String courseId, int totalNumRecipients, FeedbackResponseDetails existingResponseDetails) {
         return Templates.populateTemplate(
                 FormTemplates.TEXT_SUBMISSION_FORM,
-                Slots.DISABLED, sessionIsOpen ? "" : "disabled",
+                Slots.IS_SESSION_OPEN, sessionIsOpen ? "true" : "false",
```
Comment (HirdayGupta) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackTextQuestionDetails.java]: Since the `readonly` value needs to be set to `false` if the session is open and vice versa, I wonder if you agree that it would be better if you passed the data down as `IS_SESSION_CLOSED` instead. So your new line would read something like:
```Java
Slots.IS_SESSION_CLOSED, !sessionIsOpen ? "true" : "false"
```

Note: you would have to rename the Slots constant and its content accordingly.

File: src/main/java/teammates/common/datatransfer/questions/FeedbackTextQuestionDetails.java [hunk 3]
```diff
@@ -72,7 +72,7 @@ public String getQuestionWithoutExistingResponseSubmissionFormHtml(
             boolean sessionIsOpen, int qnIdx, int responseIdx, String courseId, int totalNumRecipients) {
         return Templates.populateTemplate(
                 FormTemplates.TEXT_SUBMISSION_FORM,
-                Slots.DISABLED, sessionIsOpen ? "" : "disabled",
+                Slots.IS_SESSION_OPEN, sessionIsOpen ? "true" : "false",
```
Comment (HirdayGupta) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackTextQuestionDetails.java]: Same here.

**Comment by HirdayGupta** [comment]
@amarlearning I'm a little new to reviewing so I'm not sure if my comments on each line were clear 😅  

So just to clarify, I was suggesting two things.

- [Changing the data retrieval from DOM in JS to the jQuery approach](https://github.com/TEAMMATES/teammates/pull/7811#discussion_r128953099) (to be consistent with how it is done in other parts of the system) 
- I was also thinking if it would be better to pass down `isSessionClosed` data to the DOM instead of `isSessionOpen` because `readonly` needs to be `true` only if session is closed. This would allow you to simply read the data using the jQuery approach and set `readonly`to the constant you store it in, making the whole code more concise and easily readable.

File: src/test/resources/pages/instructorFeedbackSubmitPageAwaiting.html [hunk 4]
```diff
@@ -258,7 +258,7 @@
               </span>
             </div>
             <div class="col-sm-10">
-              <div class="panel panel-default panel-body mce-content-body content-editor empty" data-is-session-open="false" data-length-text-id="responseLength-1-1" data-recommended-text="0" id="responsetext-1-1" name="responsetext-1-1" spellcheck="false">
+              <div class="panel panel-default panel-body mce-content-body content-editor empty" contenteditable="true" data-is-session-closed="true" data-length-text-id="responseLength-1-1" data-recommended-text="0" id="responsetext-1-1" name="responsetext-1-1" spellcheck="false" style="position: relative;">
```
Comment (HirdayGupta) [inline src/test/resources/pages/instructorFeedbackSubmitPageAwaiting.html]: @amarlearning Just a comment in passing. Am I missing something here or does this seem off? If the session is closed, then `readonly` should be set to true, meaning `contenteditable` should NOT be true. If you also think something is off can you investigate? This seems to be the case with all test files where `data-is-session-closed="true"` and since the default behaviour of readonly is to be `false`, I believe the property is not being set to `true` in the JavaScript successfully. 

One debugging measure I would suggest is to try `console.log(isSessionClosed)` after line 966 where you set the variable and see if it shows a String or a boolean.

Comment (amarlearning) [inline src/test/resources/pages/instructorFeedbackSubmitPageAwaiting.html]: Everything is working fine, I just forgot to run CI test on Godmode to update Html files after updating Java file :sweat_smile:

Comment (HirdayGupta) [inline src/test/resources/pages/instructorFeedbackSubmitPageAwaiting.html]: Hahaha I see. Travis confirms btw. Go ahead with god mode whenever you're ready :)

File: src/main/resources/feedbackQuestionTextSubmissionFormTemplate.html [hunk 5]
```diff
@@ -1,7 +1,5 @@
 <div class="panel panel-default panel-body"
-    ${disabled} id="${Const.ParamsNames.FEEDBACK_RESPONSE_TEXT}-${questionIndex}-${responseIndex}"
-    name="${Const.ParamsNames.FEEDBACK_RESPONSE_TEXT}-${questionIndex}-${responseIndex}"
-    data-length-text-id="responseLength-${questionIndex}-${responseIndex}" data-recommended-text="${recommendedLength}">
+    data-is-session-closed="${isSessionClosed}" id="${Const.ParamsNames.FEEDBACK_RESPONSE_TEXT}-${questionIndex}-${responseIndex}" name="${Const.ParamsNames.FEEDBACK_RESPONSE_TEXT}-${questionIndex}-${responseIndex}" data-length-text-id="responseLength-${questionIndex}-${responseIndex}" data-recommended-text="${recommendedLength}">
```
Comment (HirdayGupta) [inline src/main/resources/feedbackQuestionTextSubmissionFormTemplate.html]: An unrelated change of putting all the attributes of the div on the same line seems to have taken place here. Leave the original indentation and splitting of attribtues over multiple lines as it is. Only insert your attribute on one of the lines. See example of what diff should look like below: (as opposed to what it is now)

```diff
<div class="panel panel-default panel-body"
-    ${disabled} id="${Const.ParamsNames.FEEDBACK_RESPONSE_TEXT}-${questionIndex}-${responseIndex}"
+   data-is-session-closed="${isSessionClosed}" id="${Const.ParamsNames.FEEDBACK_RESPONSE_TEXT}-${questionIndex}-${responseIndex}"
    name="${Const.ParamsNames.FEEDBACK_RESPONSE_TEXT}-${questionIndex}-${responseIndex}"
    data-length-text-id="responseLength-${questionIndex}-${responseIndex}" data-recommended-text="${recommendedLength}">
    ${existingResponse}
</div>
```

File: src/main/java/teammates/common/datatransfer/questions/FeedbackTextQuestionDetails.java [hunk 6]
```diff
@@ -72,7 +72,7 @@ public String getQuestionWithoutExistingResponseSubmissionFormHtml(
             boolean sessionIsOpen, int qnIdx, int responseIdx, String courseId, int totalNumRecipients) {
         return Templates.populateTemplate(
                 FormTemplates.TEXT_SUBMISSION_FORM,
-                Slots.DISABLED, sessionIsOpen ? "" : "disabled",
+                Slots.IS_SESSION_CLOSED, sessionIsOpen ? "false" : "true",
```
Comment (HirdayGupta) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackTextQuestionDetails.java]: Might want to change this line to the following code instead.

```Java
Slots.IS_SESSION_CLOSED, !sessionIsOpen ? "true" : "false",
```

This is again something I have seen across the system. And it does read better to have the first case in a Ternary seem like the logical outcome of a `true` evaluation of the condition. However, since this is something of a special case (as the outcomes are switched), we could ask @damithc to either weigh in himself or ask another senior dev to give their opinion.

File: src/main/resources/feedbackQuestionTextSubmissionFormTemplate.html (hunk 5 above)
Comment (HirdayGupta) [inline src/main/resources/feedbackQuestionTextSubmissionFormTemplate.html]: Note: also, don't change the order of the attributes while making the change here from what you've committed in https://github.com/TEAMMATES/teammates/pull/7811/commits/b73557e3e8bdf677ac0908cc001ff7601045b6c0 as otherwise you will have to run god mode again. 😅

File: src/main/java/teammates/common/datatransfer/questions/FeedbackTextQuestionDetails.java (hunk 6 above)
Comment (amarlearning) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackTextQuestionDetails.java]: @HirdayGupta I did change the line to this
```javascript
Slots.IS_SESSION_CLOSED, !sessionIsOpen ? "true" : "false",
```
But 2 PMD test failed to suggest me to avoid using `!`

Comment (HirdayGupta) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackTextQuestionDetails.java]: @amarlearning Oh, I see. In that case go ahead with what you've committed, I guess. Let's wait to see what @damithc sir has to say, either way. :)

Comment (damithc) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackTextQuestionDetails.java]: @HirdayGupta's suggestion reads slightly better, but we can live with the current one because it is still readable (and also to avoid the PMD violation)

File: src/main/resources/feedbackQuestionTextSubmissionFormTemplate.html (hunk 5 above)
Comment (HirdayGupta) [inline src/main/resources/feedbackQuestionTextSubmissionFormTemplate.html]: @amarlearning Apologies if I wasn't clear. When I said "ignore the above comment" I meant the one about not changing the order of attributes in case you have to rerun god mode. My [first comment](https://github.com/TEAMMATES/teammates/pull/7811#discussion_r129121116) about the unrelated change of putting all attributes on the same line still holds. That is the only change that is pending.

**Comment by amarlearning** [comment]
Ready for review

File: src/main/resources/feedbackQuestionTextSubmissionFormTemplate.html [hunk 7]
```diff
@@ -1,7 +1,9 @@
 <div class="panel panel-default panel-body"
-    ${disabled} id="${Const.ParamsNames.FEEDBACK_RESPONSE_TEXT}-${questionIndex}-${responseIndex}"
-    name="${Const.ParamsNames.FEEDBACK_RESPONSE_TEXT}-${questionIndex}-${responseIndex}"
-    data-length-text-id="responseLength-${questionIndex}-${responseIndex}" data-recommended-text="${recommendedLength}">
+     data-is-session-closed="${isSessionClosed}"
+     id="${Const.ParamsNames.FEEDBACK_RESPONSE_TEXT}-${questionIndex}-${responseIndex}"
+     name="${Const.ParamsNames.FEEDBACK_RESPONSE_TEXT}-${questionIndex}-${responseIndex}"
+     data-length-text-id="responseLength-${questionIndex}-${responseIndex}"
+     data-recommended-text="${recommendedLength}">
```
Comment (HirdayGupta) [inline src/main/resources/feedbackQuestionTextSubmissionFormTemplate.html]: @amarlearning Apoloogies for being unclear.I meant that you should revert back to the orginal alignment of attributes (from the TEAMMATES/master branch) and add your one data-attribute on one of the existing lines. See the diff I posted in https://github.com/TEAMMATES/teammates/pull/7811#discussion_r129121116.

Comment (amarlearning) [inline src/main/resources/feedbackQuestionTextSubmissionFormTemplate.html]: I got what you were saying there, but don't you think its better readable this way. 

If you still think that I should change, then I will. Let me know

Comment (HirdayGupta) [inline src/main/resources/feedbackQuestionTextSubmissionFormTemplate.html]: @damithc Advice on this? I would say that restricting the number of changes a PR makes to be strictly in keeping with the change being introduced in it is preferable.

Comment (damithc) [inline src/main/resources/feedbackQuestionTextSubmissionFormTemplate.html]: Yes, we try not to include unrelated changes in a PR. If you notice minor refactorings but doing them will make the PR diff bigger, either do it as a separate PR or create a `firstTimer` issue for it.

**Comment by amarlearning** [comment]
ready for review

**Comment by HirdayGupta** [comment]
@amarlearning Turns out we have some more work to do. If you see [this line](https://github.com/TEAMMATES/teammates/blob/master/src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6#L982) `feedbackSubmissionsEdit.es6`, you will see that the tinyMCE editor gets added to ALL divs that begin with `responsetext-`. 

The question type that we have changed up till now, the text type question, is just a subset of the type of question that's affected by this issue. 

You will have to look for all template HTML files in `src/main/resources` that contain `name="${Const.ParamsNames.FEEDBACK_RESPONSE_TEXT}-` and add in the `data-is-session-closed="${isSessionClosed}"` data attribute to all those divs and then update their corresponding populating methods in the java classes from `teammates.ui.template` to populate the placeholder.

**Comment by amarlearning** [comment]
@HirdayGupta I have already checked all type of question and their behavior before. If the session is closed then `disabled` attribute is added to every `input` option by default. It was just the `Essay Type` question which takes input via `div` and where the `disabled` attribute doesn't work. Other question such as:-

1. Numeric Question takes input using `Input[type='number']`.
2. Multiple choice (single answer) Question takes input using `Input[type='radio']`.
3. Multiple choice (multiple answer) Question takes input using `Input[type='checkbox']`.
4. Distribute Point (among options) Question takes input using `Input[type='number']`.
5. Distribute Point (among recipient) Question takes input using `Input[type='number']`.
6. Team Contribution Question takes input using `select tag`.
7. Rubric Question takes input using `Input[type='radio']`.
8. Rank (option) Question takes input using `select tag`.
9. Rank (recipient) Question takes input using `select tag`.

And in all the above **9** question types the `disabled` attribute works fine. If there is any issue that you faced then let me know.

**Comment by HirdayGupta** [comment]
@amarlearning My apologies. I was mistaken earlier. You are right. All other question types work fine it was just the essay type question that needed to be fixed. 👍

File: src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6 [hunk 8]
```diff
@@ -963,10 +963,12 @@ $(document).ready(() => {
     if (typeof richTextEditorBuilder !== 'undefined') {
         $.each(textFields, (i, textField) => {
             const id = $(textField).attr('id');
+            const isSessionClosed = $(textField).data('is-session-closed');
```
Comment (HirdayGupta) [inline src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6]: It might be best to handle the case where the `data-is-session-closed` is undefined just in case. Can change this line to 
```JavaScript
const isSessionClosed = $(textField).data('is-session-closed') || false;
```

 I checked and realised that if the data is undefined then the text field becomes uneditable. I would think that making the `readonly` property `false` as the default case (to give the benefit of doubt to the session being open) might be desirable as we don't want any case in which the student is unable to submit feedback.

That being said, lets see what @damithc sir has to say about this suggestion before you implement it.

Comment (amarlearning) [inline src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6]: I can't think of any case where `data-is-session-closed` will be undefined. we already stated in the backend that if the session is closed `data-is-session-closed = true` else `data-is-session-closed = false`. :sweat_smile: 
```java
Slots.IS_SESSION_CLOSED, sessionIsOpen ? "false" : "true"
```

Comment (HirdayGupta) [inline src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6]: @amarlearning I know that, which is why I said just in case 😅. This is something I picked up from @whipermr5. (If you're interested you can see the discussion [here](https://github.com/TEAMMATES/teammates/pull/7715#issuecomment-316595294)). Even if we know that something can never be undefined, it never hurts to be prepared for the exception.

Comment (whipermr5) [inline src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6]: Haha that discussion was more about readability. But yes! There is value to defensive coding, which is what this is :)

Comment (HirdayGupta) [inline src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6]: @whipermr5 while we wait for @damithc's input, what would be your opinion on the [above suggestion](https://github.com/TEAMMATES/teammates/pull/7811#discussion_r129743402) about defaulting to `false` if the value is undefined?

Comment (damithc) [inline src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6]: > That being said, lets see what @damithc sir has to say about this suggestion before you implement it.

I have no strong opinions about this, as I'm not a JS expert. So, I'll let the other senior devs decide 😛

Comment (whipermr5) [inline src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6]: > I would think that making the readonly property false as the default case (to give the benefit of doubt to the session being open) might be desirable as we don't want any case in which the student is unable to submit feedback.

Agree with this 👍

File: src/main/java/teammates/common/datatransfer/questions/FeedbackTextQuestionDetails.java (hunk 6 above)
Comment (whipermr5) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackTextQuestionDetails.java]: This might be even more readable:
```java
Slots.IS_SESSION_CLOSED, Boolean.toString(!sessionIsOpen)
```
However, I actually prefer if we stick to `IS_SESSION_OPEN`, which is even more readable:
```java
Slots.IS_SESSION_OPEN, Boolean.toString(sessionIsOpen)
```
The final negation can happen at the point of usage (when setting the readonly field in JavaScript). When data is still in the process of being passed around, I feel it should be made as readable as possible.

File: src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6 [hunk 9]
```diff
@@ -963,10 +963,12 @@ $(document).ready(() => {
     if (typeof richTextEditorBuilder !== 'undefined') {
         $.each(textFields, (i, textField) => {
             const id = $(textField).attr('id');
+            const isSessionClosed = $(textField).data('is-session-closed') || false;
```
Comment (whipermr5) [inline src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6]: I believe jQuery converts the case automatically:
```javascript
const isSessionClosed = $(textField).data('isSessionClosed') || false;
```
If we are going with my above suggestion, the fact that we're giving the benefit of the doubt that the session is still open becomes clearer:
```javascript
const isSessionOpen = $(textField).data('isSessionOpen') || true;
```

File: src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6 [hunk 10]
```diff
@@ -963,10 +963,12 @@ $(document).ready(() => {
...
 
             /* eslint-disable camelcase */ // The property names are determined by external library (tinymce)
             richTextEditorBuilder.initEditor(`#${id}`, {
                 inline: true,
+                readonly: isSessionClosed,
```
Comment (whipermr5) [inline src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6]: The negation can happen here:
```javascript
readonly: !isSessionOpen
```
Readers of the code would only need to think about the negation just this once.

File: src/main/java/teammates/common/datatransfer/questions/FeedbackTextQuestionDetails.java (hunk 6 above)
Comment (HirdayGupta) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackTextQuestionDetails.java]: @whipermr5 Agree with your suggestion about using the `Boolean.toString()` method. Didn't strike me at all! Thanks! 😅 

However, I don't fully agree with changing to `IS_SESSION_OPEN`. Yes, your argument about making the passing around of variables more readable so that people only have to think about the negation once makes sense. 

However, the fact that we are renaming it to `IS_SESSION_CLOSED` also means, that when readers have taken into account the negation that happens here, they only have to think about it that one time, because after that the variable name makes it self-explanatory. Keep in mind that we had to add this slot in, meaning that there is no precedent for passing `IS_SESSION_OPEN` that will make readers wonder about the departure from convention. 

Here, we are setting the precedent. And this was why I thought it would be better to pass down `IS_SESSION_CLOSED` instead. Because in the end, the truth value of that statement, "is the session closed?" is what we care about in the final output (in this case, if true, disable div). 

However, I do, of course, defer to your better judgment on this. Just thought I should make one final case for my suggestion. :)

File: src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6 (hunk 9 above)
Comment (HirdayGupta) [inline src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6]: @whipermr5 Just as a continuation to the [point I made above](https://github.com/TEAMMATES/teammates/pull/7811#discussion_r131247789), I think the variable name makes it clear enough about which side is receiving the benefit of the doubt either way.

Comment (amarlearning) [inline src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6]: > ```javascript
> const isSessionOpen = $(textField).data('isSessionOpen') || true;
> ```

@whipermr5 we can't use `true` as the second value because in pipe operator

- If the first value is `false`, it checks the second value. If it's `true`, it returns `true` and if it's `false`, it returns `false`.

- If the first value is `true`, it always returns `true`, no matter what the second value is.

***

Let us consider our case in which if the `$(textField).data('isSessionOpen')` value is `false` that means the session is closed. But still the `const isSessionOpen` value will evaluate to `true` because of this :point_down:

```javascript
const isSessionOpen = $(textField).data('isSessionOpen') || true;
```
_Set `$(textField).data('isSessionOpen')` value to `false`_

```javascript
const isSessionOpen = false || true;
```
Now that we have `false` as our first value and `true` as our second value, the `||` operator will evaluate to `true` which means that the `const isSessionOpen` value will be `true` which should have been `false`.

Not very good in explaining, but I hope you got this :sweat_smile:

Comment (whipermr5) [inline src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6]: @amarlearning Oops!! 😅😅 Thanks for catching that!

Comment (whipermr5) [inline src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6]: > I think the variable name makes it clear enough about which side is receiving the benefit of the doubt either way.

@HirdayGupta `isSessionClosed = false` is a double negation and slightly less clear haha.

Comment (whipermr5) [inline src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6]: As for the setting of `isSessionOpen`, maybe use this?
```javascript
const isSessionOpenData = $(textField).data('isSessionOpen');
const isSessionOpen = typeof(isSessionOpenData) === 'boolean' ? isSessionOpenData : true;
```

File: src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6 (hunk 10 above)
Comment (whipermr5) [inline src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6]: The main reason I don't like this is that even when directly assigning `readonly` to `isSessionClosed`, the reader would still have to pause to think why the two are equivalent. Not sure if you agree with me, but I'd argue that most readers still go through this thought process:
```javascript
readonly: isSessionClosed ? true : false,
```
If `readonly` is instead assigned to `!isSessionOpen`, the thought process would be more like:
```javascript
readonly: isSessionOpen ? false : true,
```
In fact, the above might be the most readable.

Comment (HirdayGupta) [inline src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6]: @whipermr5 I understand your reservations about passing `IS_SESSION_CLOSED` and I agree that your last statement would be the most readable. @amarlearning Go ahead with @whipermr5's suggestions. I have nothing more to add :)

**Comment by amarlearning** [comment]
sorry for the delay. ready for review

**Comment by amarlearning** [comment]
@whipermr5 reminder :sweat_smile:
